### ISSUE TITLE
A2 bootloader stuck on booting from Flash

### ISSUE THREAD

**Issue by abdullahyildiz** [openhwgroup/core-v-mcu-cli-test#47]
We applied the [instructions](https://github.com/openhwgroup/core-v-mcu-cli-test/tree/main/cli_test#programming-qspi-flash) to program QSPI Flash (32MB serial NOR Flash memory, Pmod SF3) and then verified that Flash has been programmed (we used `flash_peek` command in CLI test to read a set of addresses in Flash memory). However, A2 Bootloader does not complete the load operation and seems to have stuck:

```
Jan 28 2022 14:42:41
A2 Bootloader Bootsel=1 rr=00
Loading Section 00000000 to ffff080
```

There could be many reasons causing this but I suspect that I use a wrong version of the A2 bootloader. Does the code in main branch correspond to stable version?

**Comment by MikeOpenHWGroup** [comment]
> Does the code in main branch correspond to stable version?

Yes, that is my understanding.  Having said that, I have not attempted to program the QSPI.  @gmartin102 can you offer any insight?

**Comment by gmartin102** [comment]
The A2 Bootloader assumes a Header in the SPI flash to indicate where the data in SPI resides, where in memory it should be loader, how much to load and where to begin execution in memory. The cli-test repository has a file header.bin that is a hand crafted header.   The cli_test.bin is prepended to the header.bin to create the data to be loaded in the SPI.  The header.bin contains a 16 byte header structure and 16 memory area structures as defined in the A2_boot flash.h file.

typedef struct {
	uint32_t nextDesc;  // set to 0x10 the next open space in SPI -- 
	uint32_t nbAreas;   // set to 0x1 indicating only 1 memory area to be loaded
	uint32_t entry;       // set to 0x1c000880 -- the start address of the cli_test program as specified in the linker file
	uint32_t bootaddr; // set to 0x1c000880
} flash_v2_header_t;

There are 16 of these structures only the first one is 
typedef struct {
	uint32_t start;  // 0x110 address in SPI flash of data to be loaded
	uint32_t ptr;     // 0x1c000800 address in memory to load data to 
	uint32_t size;   // 0x30000.  copy 192 bytes of data from spy to memory This should be the size of the section
	uint32_t blocks; // 0x30 number of 1K blocks to copy
} flash_v2_mem_area_t;
